File: 01_Arrays_&_Hashing/02_Medium/EncodeAndDecodeStrings.py
class Solution:
    '''
    @param: strs: a list of strings
    @return: encodes a list of strings to a single string
    '''

    def encode(self, strs):
        res = ""
        for s in strs:
            res += str(len(s)) + "#" + s
        
        return res 
    
    '''
    @param: s: A string
    @return: decodes a single string into a list of strings
    '''

    def decode(self, s):
        res, i = [], 0

        while i < len(s):
            j = 0 
            while s[j] != "#":
                j += 1
            length = int(s[i:j])
            res.append(s[j+1: j+1+length])
            i = j + 1 + length
        
        return res 

# store length of str before each string and delimiter like '#';
# Space | Time: O(n) | O(n)

File: 01_Arrays_&_Hashing/02_Medium/C++/EncodeAndDecodeStrings.cpp
#include <string>
#include <vector>

class Code {
    public:
        std::string encode(std::vector<std::string>& strs){
            std::string res = "";
            for (const string& word : strs){
                res += to_string(word.size()) + "#" + word;
            }
            return res;
        };

        std::vector<std::string> decode(std::string& s){
            std::vector<std::string> res;
            int i = 0;

            while (i < s.size()){
                int j = i;
                while (s[j] != '#'){
                    j++;
                }
                int length = stoi(s.substr(i, j - i));
                res.push_back(s.substr(j+1, length));
                i = j + 1 + length;
            }
            return res;
        };
};

// Store the length of each string before a delimiter like "#".
// Time | Space : O(n) | O(1)
